Makefile

```diff
@@ -0,0 +1,31 @@
+# define the name of the virtual environment directory
+VENV := venv
+
+# default target, when make executed without arguments
+all: venv
+
+$(VENV)/bin/activate: requirements.txt
+	python3 -m venv $(VENV)
+	./$(VENV)/bin/pip install -r requirements.txt
+	./$(VENV)/bin/pip install pandas
+
+# venv is a shortcut target
+venv: $(VENV)/bin/activate
+
+clean:
+	rm -rf $(VENV)
+	find . -type f -name '*.pyc' -delete
+
+run-clickhouse:
+	@docker run -p 9000:9000 --ulimit nofile=262144:262144 -it clickhouse/clickhouse-server
+
+unit: venv
+	./$(VENV)/bin/python3 -m unittest discover -s tests/unit
+
+integration: venv
+	./$(VENV)/bin/python3 -m unittest discover -s tests/integration
+
+tests: venv
+	./$(VENV)/bin/python3 -m unittest discover -s tests
+
+.PHONY: all venv unit clean
```

Comment:
What does this mean?

Reply:
It's a common practise and it has two goals:
- better performance (because make is a file-oriented utility and it tries to find target file by default.
- exclude incorrect behaviour (if we create file `clean`, make skips clean command)

You can get additional information from the following link:
https://www.gnu.org/software/make/manual/html_node/Phony-Targets.html

Also I've prepared small example for you:

no phony:
<img width="218" alt="image" src="https://user-images.githubusercontent.com/1254188/201906978-83e91a88-f8ba-4220-a7c8-30932d379d69.png">

with phony:
<img width="218" alt="image" src="https://user-images.githubusercontent.com/1254188/201907126-e145adc4-18a7-4525-bb9c-37aa7bbf45a7.png">

Reply:
Very clear, thank you! I will add it to my bookmarks 😉